File: Assignment 3 /task 1 and task 2.1/task2_1.scala
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import java.io.{File, PrintWriter}
import scala.collection.mutable
import scala.math._
import scala.util.{Try, Success, Failure}

object task2_1 {
  def main(args: Array[String]): Unit = {
    // Check command line arguments
    if (args.length != 3) {
      println("Usage: task2_1 <train_file_name> <test_file_name> <output_file_name>")
      System.exit(1)
    }

    val trainFile = args(0)
    val testFile = args(1)
    val outputFile = args(2)

    // Initialize Spark
    val conf = new SparkConf().setAppName("Task2_1_ItemBasedCF")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    // Track execution time
    val startTime = System.currentTimeMillis()

    // Read training data
    val trainRDD = readTrainData(sc, trainFile)
    trainRDD.cache()

    // Compute averages
    val (businessAvg, userAvg, globalAvg) = computeAverages(trainRDD)

    // Build rating dictionaries
    val (userBusinesses, businessUsers) = buildRatingDicts(trainRDD)

    // Read test data
    val testPairs = readTestData(sc, testFile)

    // Initialize similarity cache
    val similarityCache = mutable.Map[(String, String), Double]()

    // Make predictions
    val predictions = testPairs.map { case (userId, businessId) =>
      val predictedRating = predictRating(
        userId, businessId, userBusinesses, businessUsers,
        businessAvg, userAvg, globalAvg, similarityCache
      )
      (userId, businessId, predictedRating)
    }

    // Write predictions to output file
    val pw = new PrintWriter(new File(outputFile))
    try {
      pw.println("user_id,business_id,prediction")
      predictions.foreach { case (userId, businessId, rating) =>
        pw.println(s"$userId,$businessId,$rating")
      }
    } finally {
      pw.close()
    }

    // Try to calculate RMSE if validation data has ratings (for self-evaluation)
    try {
      // Read validation data (test file) with ratings
      val validationLines = sc.textFile(testFile)
      val validationHeader = validationLines.first()
      
      val validationRDD = validationLines
        .filter(_ != validationHeader)  // Skip header
        .map(_.split(","))
        .filter(r => r.length >= 3)  // Make sure rating column exists
        .map(r => ((r(0), r(1)), r(2).toDouble))
      
      // Check if we have validation data with ratings
      if (!validationRDD.isEmpty()) {
        // Convert to map for faster lookup
        val validationDict = validationRDD.collectAsMap()
        
        // Initialize error variables
        var totalSquaredError = 0.0
        var count = 0
        
        // Initialize counters for different error ranges
        val errorRanges = mutable.Map(
          ">=0 and <1" -> 0,
          ">=1 and <2" -> 0,
          ">=2 and <3" -> 0,
          ">=3 and <4" -> 0,
          ">=4" -> 0
        )
        
        // Calculate error for each prediction
        predictions.foreach { case (userId, businessId, predRating) =>
          val key = (userId, businessId)
          if (validationDict.contains(key)) {
            val actual = validationDict(key)
            val error = abs(predRating - actual)
            
            // Update RMSE calculation
            totalSquaredError += error * error
            count += 1
            
            // Categorize error into appropriate range
            if (error < 1) errorRanges(">=0 and <1") += 1
            else if (error < 2) errorRanges(">=1 and <2") += 1
            else if (error < 3) errorRanges(">=2 and <3") += 1
            else if (error < 4) errorRanges(">=3 and <4") += 1
            else errorRanges(">=4") += 1
          }
        }
        
        // Calculate and print RMSE
        if (count > 0) {
          val rmse = sqrt(totalSquaredError / count)
          println(s"RMSE: $rmse")
        }
        
        // Print error distribution
        println("Error Distribution:")
        errorRanges.foreach { case (rangeLabel, count) =>
          println(s"  $rangeLabel: $count")
        }
      }
    } catch {
      case e: Exception => 
        // If RMSE calculation fails, just skip it
        println("Skipping RMSE calculation (expected if test file doesn't have ratings)")
    }

    val executionTime = (System.currentTimeMillis() - startTime) / 1000.0
    println(s"Task 2.1 completed successfully in $executionTime seconds.")
    sc.stop()
  }

  def readTrainData(sc: SparkContext, trainFile: String): RDD[(String, String, Double)] = {
    // Load dataset
    val lines = sc.textFile(trainFile)
    val header = lines.first()

    // Parse training data
    lines
      .filter(_ != header)
      .map(_.split(","))
      .filter(r => r.length >= 3)
      .map(r => (r(0), r(1), r(2).toDouble))
  }

  def readTestData(sc: SparkContext, testFile: String): Array[(String, String)] = {
    // Load dataset
    val lines = sc.textFile(testFile)
    val header = lines.first()

    // Parse test data without ratings
    lines
      .filter(_ != header)
      .map(_.split(","))
      .filter(r => r.length >= 2)
      .map(r => (r(0), r(1)))
      .collect()
  }

  def computeAverages(trainRDD: RDD[(String, String, Double)]): (Map[String, Double], Map[String, Double], Double) = {
    // Business averages
    val businessAvg = trainRDD
      .map(x => (x._2, (x._3, 1)))
      .reduceByKey((a, b) => (a._1 + b._1, a._2 + b._2))
      .mapValues(t => t._1 / t._2)
      .collectAsMap()
      .toMap

    // User averages
    val userAvg = trainRDD
      .map(x => (x._1, (x._3, 1)))
      .reduceByKey((a, b) => (a._1 + b._1, a._2 + b._2))
      .mapValues(t => t._1 / t._2)
      .collectAsMap()
      .toMap

    // Global average
    val (total, count) = trainRDD
      .map(x => (x._3, 1))
      .reduce((a, b) => (a._1 + b._1, a._2 + b._2))
    val globalAvg = total / count

    (businessAvg, userAvg, globalAvg)
  }

  def buildRatingDicts(trainRDD: RDD[(String, String, Double)]): (Map[String, Map[String, Double]], Map[String, Map[String, Double]]) = {
    // User -> Businesses dictionary
    val userBusinesses = trainRDD
      .map(x => (x._1, (x._2, x._3)))
      .groupByKey()
      .mapValues(recs => recs.toMap)
      .collectAsMap()
      .toMap

    // Business -> Users dictionary
    val businessUsers = trainRDD
      .map(x => (x._2, (x._1, x._3)))
      .groupByKey()
      .mapValues(recs => recs.toMap)
      .collectAsMap()
      .toMap

    (userBusinesses, businessUsers)
  }

  def computePearsonSimilarity(
    business1: String,
    business2: String,
    businessUsers: Map[String, Map[String, Double]],
    businessAvg: Map[String, Double],
    similarityCache: mutable.Map[(String, String), Double]
  ): Double = {
    // Use cache for faster computation
    val cacheKey = if (business1 < business2) (business1, business2) else (business2, business1)
    
    // Return cached value if available
    if (similarityCache.contains(cacheKey)) {
      return similarityCache(cacheKey)
    }

    // Find common users
    val users1 = businessUsers(business1).keySet
    val users2 = businessUsers(business2).keySet
    val commonUsers = users1.intersect(users2)

    // Handle special cases with few common users
    if (commonUsers.size <= 1) {
      val similarity = (5.0 - abs(businessAvg(business1) - businessAvg(business2))) / 5.0
      similarityCache(cacheKey) = similarity
      return similarity
    }

    // For exactly 2 common users, use average of rating differences
    if (commonUsers.size == 2) {
      val usersList = commonUsers.toList
      val sim1 = (5.0 - abs(businessUsers(business1)(usersList(0)) - businessUsers(business2)(usersList(0)))) / 5.0
      val sim2 = (5.0 - abs(businessUsers(business1)(usersList(1)) - businessUsers(business2)(usersList(1)))) / 5.0
      val similarity = (sim1 + sim2) / 2
      similarityCache(cacheKey) = similarity
      return similarity
    }

    // Extract ratings for common users
    val ratings1 = commonUsers.toList.map(user => businessUsers(business1)(user))
    val ratings2 = commonUsers.toList.map(user => businessUsers(business2)(user))

    // Calculate means
    val mean1 = ratings1.sum / ratings1.size
    val mean2 = ratings2.sum / ratings2.size

    // Center the ratings
    val centered1 = ratings1.map(_ - mean1)
    val centered2 = ratings2.map(_ - mean2)

    // Calculate Pearson correlation
    val numerator = centered1.zip(centered2).map { case (a, b) => a * b }.sum
    val denominator = sqrt(centered1.map(a => a * a).sum) * sqrt(centered2.map(b => b * b).sum)

    val similarity = if (denominator == 0) 0.0 else numerator / denominator

    // Cache the result
    similarityCache(cacheKey) = similarity
    similarity
  }

  def predictRating(
    userId: String,
    businessId: String,
    userBusinesses: Map[String, Map[String, Double]],
    businessUsers: Map[String, Map[String, Double]],
    businessAvg: Map[String, Double],
    userAvg: Map[String, Double],
    globalAvg: Double,
    similarityCache: mutable.Map[(String, String), Double],
    neighborCount: Int = 15
  ): Double = {
    // Check if user has already rated this business
    if (userBusinesses.contains(userId) && userBusinesses(userId).contains(businessId)) {
      return userBusinesses(userId)(businessId)
    }

    // Handle cold start problems
    if (!userBusinesses.contains(userId)) {
      return businessAvg.getOrElse(businessId, globalAvg)
    }

    if (!businessUsers.contains(businessId)) {
      return userAvg.getOrElse(userId, globalAvg)
    }

    // Find similar items that the user has rated
    val similarItems = userBusinesses(userId).flatMap { case (otherBusiness, rating) =>
      // Skip if same business or not in training data
      if (otherBusiness == businessId || !businessUsers.contains(otherBusiness)) {
        None
      } else {
        // Compute similarity
        val similarity = computePearsonSimilarity(
          businessId, otherBusiness, businessUsers, businessAvg, similarityCache
        )
        
        // Only use positive similarities
        if (similarity > 0) Some((similarity, rating)) else None
      }
    }.toList

    // Fallback if no similar items found
    if (similarItems.isEmpty) {
      return userAvg.getOrElse(userId, globalAvg)
    }

    // Calculate weighted average
    val topNeighbors = similarItems.sortBy(-_._1).take(neighborCount)
    val numerator = topNeighbors.map { case (sim, r) => sim * r }.sum
    val denominator = topNeighbors.map { case (sim, _) => abs(sim) }.sum

    // Handle edge case
    if (denominator == 0) {
      return userAvg.getOrElse(userId, globalAvg)
    }

    // Final rating prediction
    val predictedRating = numerator / denominator
    min(max(predictedRating, 1.0), 5.0)
  }
}

File: Assignment 3 /task 1 and task 2.1/task2_1.py
import sys
import time
import csv
import math
from itertools import combinations
from collections import defaultdict
from pyspark import SparkContext, SparkConf

# ===================================================================
# HELPER FUNCTIONS FOR DATA PROCESSING
# ===================================================================

def read_train_data(train_file):
    """
    Reads the training CSV file and processes it into (user, business, rating) records.
    Handles potential CSV format issues to prevent index errors.
    
    Returns:
        RDD of tuples (user_id, business_id, rating)
    """
    # Load the raw CSV file as an RDD of text lines
    raw_data = sc.textFile(train_file)
    
    # Extract the header line (first line of the CSV)
    header = raw_data.first()
    
    # Filter out header and safely parse rows with these steps:
    # 1. Remove the header row
    # 2. Split each row by comma delimiter
    # 3. Filter out any malformed rows that don't have enough columns
    # 4. Convert to the required format with rating as float
    return (raw_data.filter(lambda row: row != header)
            .map(lambda row: row.split(","))
            .filter(lambda r: len(r) >= 3)  # Filter out malformed rows
            .map(lambda r: (r[0], r[1], float(r[2]))))

def read_test_data(test_file, include_ratings=False):
    """
    Reads the test CSV file containing (user, business) pairs.
    Handles potential CSV format issues to prevent index errors.
    
    Args:
        test_file (str): Path to test file
        include_ratings (bool): Whether to include ratings in the result or just user-business pairs
    
    Returns:
        RDD of tuples (user_id, business_id) or (user_id, business_id, rating)
    """
    # Load the raw CSV file as an RDD of text lines
    raw_data = sc.textFile(test_file)
    
    # Extract the header line (first line of the CSV)
    header = raw_data.first()
    
    # Split rows and filter out malformed ones:
    # 1. Remove the header row
    # 2. Split each row by comma delimiter
    # 3. Apply different filter based on whether we need ratings or not
    parsed_data = (raw_data.filter(lambda row: row != header)
                   .map(lambda row: row.split(","))
                   .filter(lambda r: len(r) >= 3 if include_ratings else len(r) >= 2))
    
    # Return data in the appropriate format:
    # - If ratings are needed, include the third column as a float
    # - Otherwise, just return user_id and business_id
    if include_ratings:
        return parsed_data.map(lambda r: (r[0], r[1], float(r[2])))
    else:
        return parsed_data.map(lambda r: (r[0], r[1]))

def compute_averages(train_rdd):
    """
    Computes average ratings at three levels:
    - Per business: a mapping of business -> average rating.
    - Per user: a mapping of user -> average rating.
    - Global average: overall average rating.
    
    Returns:
        Tuple: (business_avg, user_avg, global_avg)
    """
    # ---------------------------
    # 1. Calculate business averages
    # ---------------------------
    # For each record, create (business_id, (rating, 1)) pairs
    # Reduce by key to sum ratings and count per business
    # Compute average by dividing sum by count
    business_avg = (train_rdd.map(lambda x: (x[1], (x[2], 1)))
                              .reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))
                              .mapValues(lambda t: t[0] / t[1])
                              .collectAsMap())
    
    # ---------------------------
    # 2. Calculate user averages
    # ---------------------------
    # Similar approach but grouping by user_id instead
    user_avg = (train_rdd.map(lambda x: (x[0], (x[2], 1)))
                          .reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))
                          .mapValues(lambda t: t[0] / t[1])
                          .collectAsMap())
    
    # ---------------------------
    # 3. Calculate global average
    # ---------------------------
    # Extract all ratings with a count of 1, then reduce to get total sum and count
    # Divide to get the overall average rating across all users and businesses
    total, count = train_rdd.map(lambda x: (x[2], 1)).reduce(lambda a, b: (a[0] + b[0], a[1] + b[1]))
    global_avg = total / count
    
    return business_avg, user_avg, global_avg

def build_rating_dicts(train_rdd):
    """
    Builds two key dictionaries from the training data:
    1. user_businesses: mapping user -> {business: rating}
    2. business_users: mapping business -> {user: rating}
    
    These dictionaries provide fast lookup for ratings during prediction and
    similarity computation.
    
    Returns:
        Tuple: (user_businesses, business_users)
    """
    # ---------------------------
    # 1. Build user -> businesses dictionary
    # ---------------------------
    # For each record, create (user_id, (business_id, rating)) pairs
    # Group by user_id to get all businesses rated by each user
    # Convert grouped values to dictionaries for fast lookup
    user_businesses = (train_rdd.map(lambda x: (x[0], (x[1], x[2])))
                           .groupByKey()
                           .mapValues(lambda recs: dict(recs))
                           .collectAsMap())
    
    # ---------------------------
    # 2. Build business -> users dictionary
    # ---------------------------
    # Similar approach but in reverse: group by business_id
    business_users = (train_rdd.map(lambda x: (x[1], (x[0], x[2])))
                               .groupByKey()
                               .mapValues(lambda recs: dict(recs))
                               .collectAsMap())
    
    return user_businesses, business_users

# ===================================================================
# SIMILARITY CALCULATION FUNCTIONS
# ===================================================================

def compute_pearson_similarity(business_1, business_2, business_users, business_avg, similarity_cache):
    """
    Computes the Pearson similarity between two businesses.
    Uses a cache to avoid redundant computations.
    
    Returns:
        float: The Pearson similarity between the two businesses
    """
    # ---------------------------
    # 1. Check cache for existing calculation
    # ---------------------------
    # Create a consistent key for the cache by sorting business IDs
    # This ensures (A,B) and (B,A) use the same cache entry
    cache_key = tuple(sorted([business_1, business_2]))
    
    # Return cached value if already computed to save time
    if cache_key in similarity_cache:
        return similarity_cache[cache_key]
    
    # ---------------------------
    # 2. Find common users who rated both businesses
    # ---------------------------
    # Get the set of users who rated each business
    users1 = set(business_users[business_1].keys())
    users2 = set(business_users[business_2].keys())
    
    # Find the intersection (users who rated both businesses)
    common_users = users1.intersection(users2)
    
    # ---------------------------
    # 3. Handle special cases with few common users
    # ---------------------------
    # If <= 1 common user, use fallback similarity based on average rating difference
    if len(common_users) <= 1:
        similarity = (5.0 - abs(business_avg[business_1] - business_avg[business_2])) / 5.0
        similarity_cache[cache_key] = similarity
        return similarity
    
    # Optimization for exactly 2 common users: use average of rating differences
    if len(common_users) == 2:
        users_list = list(common_users)
        sim1 = (5.0 - abs(business_users[business_1][users_list[0]] - 
                         business_users[business_2][users_list[0]])) / 5.0
        sim2 = (5.0 - abs(business_users[business_1][users_list[1]] - 
                         business_users[business_2][users_list[1]])) / 5.0
        similarity = (sim1 + sim2) / 2
        similarity_cache[cache_key] = similarity
        return similarity
    
    # ---------------------------
    # 4. Calculate Pearson correlation for 3+ common users
    # ---------------------------
    # Extract ratings for common users
    ratings1 = [business_users[business_1][user] for user in common_users]
    ratings2 = [business_users[business_2][user] for user in common_users]
    
    # Calculate means for both sets of ratings
    mean1 = sum(ratings1) / len(ratings1)
    mean2 = sum(ratings2) / len(ratings2)
    
    # Center the ratings by subtracting means (creates zero-centered data)
    centered1 = [r - mean1 for r in ratings1]
    centered2 = [r - mean2 for r in ratings2]
    
    # Calculate numerator: sum of (x-mean_x)*(y-mean_y)
    numerator = sum(a * b for a, b in zip(centered1, centered2))
    
    # Calculate denominator: sqrt(sum((x-mean_x)²) * sum((y-mean_y)²))
    denominator = (sum(a ** 2 for a in centered1) ** 0.5) * (sum(b ** 2 for b in centered2) ** 0.5)
    
    # Handle division by zero case
    if denominator == 0:
        similarity = 0.0
    else:
        similarity = numerator / denominator
    
    # Cache the result for future use and return
    similarity_cache[cache_key] = similarity
    return similarity

# ===================================================================
# RATING PREDICTION FUNCTION
# ===================================================================

def predict_rating(user_id, business_id, user_businesses, business_users, 
                  business_avg, user_avg, global_avg, similarity_cache, neighbor_count=15):
    """
    Predicts the rating for a given (user, business) pair using item-based CF.
    
    Returns:
        float: Predicted rating (1.0-5.0)
    """
    # ---------------------------
    # 1. Check if user has already rated this business
    # ---------------------------
    # If so, return the actual rating (though this is unlikely in test data)
    if user_id in user_businesses and business_id in user_businesses[user_id]:
        return user_businesses[user_id][business_id]
    
    # ---------------------------
    # 2. Handle cold start problems (new user or new business)
    # ---------------------------
    # Case 2a: New user - return business average or global average
    if user_id not in user_businesses:
        return business_avg.get(business_id, global_avg)
    
    # Case 2b: New business - return user average or global average
    if business_id not in business_users:
        return user_avg.get(user_id, global_avg)
    
    # ---------------------------
    # 3. Use item-based CF with Pearson similarity
    # ---------------------------
    # Find similar businesses that the user has rated
    similar_items = []
    
    # For each business the user has rated, compute similarity with target business
    for other_business, rating in user_businesses[user_id].items():
        # Skip if same business or if other business not in training data
        if other_business == business_id or other_business not in business_users:
            continue
        
        # Compute similarity between the target and other business
        similarity = compute_pearson_similarity(
            business_id, other_business, business_users, business_avg, similarity_cache
        )
        
        # Only use positive similarities (more informative for prediction)
        if similarity > 0:
            similar_items.append((similarity, rating))
    
    # ---------------------------
    # 4. Fallback if no similar items found
    # ---------------------------
    if not similar_items:
        return user_avg.get(user_id, global_avg)
    
    # ---------------------------
    # 5. Calculate weighted average of ratings from similar items
    # ---------------------------
    # Sort by similarity (descending) and select top neighbors
    similar_items.sort(key=lambda x: -x[0])
    top_neighbors = similar_items[:neighbor_count]
    
    # Compute weighted average rating (sum of similarity * rating / sum of similarities)
    numerator = sum(sim * r for sim, r in top_neighbors)
    denominator = sum(abs(sim) for sim, r in top_neighbors)
    
    # Handle edge case of zero denominator
    if denominator == 0:
        return user_avg.get(user_id, global_avg)
    
    # Calculate final rating and ensure it's within valid range [1.0, 5.0]
    predicted_rating = numerator / denominator
    return min(max(predicted_rating, 1.0), 5.0)

# ===================================================================
# EVALUATION FUNCTIONS
# ===================================================================

def calculate_rmse(predictions, validation_rdd):
    """
    Calculates the Root Mean Square Error by comparing predictions against validation data.
    
    Args:
        predictions (list): List of (user_id, business_id, predicted_rating) tuples
        validation_rdd (RDD): RDD of validation data with (user_id, business_id, actual_rating)
        
    Returns:
        float: RMSE value
    """
    # ---------------------------
    # 1. Create fast lookup dictionary for predictions
    # ---------------------------
    # Convert predictions list to dictionary with (user_id, business_id) as key
    pred_dict = {(user_id, business_id): rating for user_id, business_id, rating in predictions}
    
    # ---------------------------
    # 2. Map validation data to (predicted, actual) pairs
    # ---------------------------
    def get_prediction(item):
        user_id, business_id, actual = item
        key = (user_id, business_id)
        if key in pred_dict:
            return (pred_dict[key], actual)
        return None
    
    # Apply the mapping function and filter out any missing predictions
    paired_ratings = validation_rdd.map(get_prediction).filter(lambda x: x is not None)
    
    # ---------------------------
    # 3. Calculate squared errors
    # ---------------------------
    # For each (predicted, actual) pair, compute the squared difference
    squared_errors = paired_ratings.map(lambda x: (x[0] - x[1]) ** 2)
    
    # ---------------------------
    # 4. Compute final RMSE value
    # ---------------------------
    # Count the number of pairs (needed for denominator)
    count = squared_errors.count()
    
    # Handle edge case of no valid predictions
    if count == 0:
        return 0.0
    
    # Sum all squared errors and calculate RMSE
    total_error = squared_errors.sum()
    return math.sqrt(total_error / count)

# ===================================================================
# MAIN FUNCTION
# ===================================================================

def main():
    """
    Main function for Task 2_1: Item-based Collaborative Filtering
    """
    # ---------------------------
    # 1. Setup and Initialization
    # ---------------------------
    # Start timing for performance measurement
    start_time = time.time()
    
    # Initialize Spark context with appropriate configuration
    conf = SparkConf().setAppName("Task2_1_ItemBasedCF")
    global sc
    sc = SparkContext(conf=conf)
    sc.setLogLevel("ERROR")  # Reduce log verbosity
    
    # Parse command line arguments
    if len(sys.argv) != 4:
        print("Usage: spark-submit task2_1.py <train_file_name> <test_file_name> <output_file_name>")
        sys.exit(1)
    
    train_file = sys.argv[1]
    test_file = sys.argv[2]
    output_file = sys.argv[3]
    
    # ---------------------------
    # 2. Data Loading and Preprocessing
    # ---------------------------
    # Read training data and cache for multiple uses
    train_rdd = read_train_data(train_file)
    train_rdd.cache()  # Important optimization since we'll use this RDD multiple times
    
    # Compute rating averages at different levels (business, user, global)
    # These will be used for cold-start handling and similarity calculation
    business_avg, user_avg, global_avg = compute_averages(train_rdd)
    
    # Build dictionaries for fast rating lookups:
    # - user_businesses: map users to businesses they've rated
    # - business_users: map businesses to users who rated them
    user_businesses, business_users = build_rating_dicts(train_rdd)
    
    # ---------------------------
    # 3. Read Test Data (Prediction Targets)
    # ---------------------------
    # Read test data without ratings - just (user_id, business_id) pairs
    # We'll predict ratings for these pairs
    test_rdd = read_test_data(test_file, include_ratings=False)
    test_pairs = test_rdd.collect()  # Collect to driver for prediction
    
    # ---------------------------
    # 4. Similarity Cache Initialization
    # ---------------------------
    # Create an empty cache for storing computed similarities
    # This significantly speeds up predictions by avoiding redundant calculations
    similarity_cache = {}
    
    # ---------------------------
    # 5. Prediction Generation
    # ---------------------------
    # Make predictions for each user-business pair in the test set
    predictions = []
    for user_id, business_id in test_pairs:
        # Use item-based collaborative filtering to predict rating
        predicted_rating = predict_rating(
            user_id, business_id, user_businesses, business_users,
            business_avg, user_avg, global_avg, similarity_cache
        )
        predictions.append((user_id, business_id, predicted_rating))
    
    # ---------------------------
    # 6. Result Sorting and Output
    # ---------------------------
    # Sort results lexicographically as required by assignment
    predictions.sort(key=lambda x: (x[0], x[1]))
    
    # Write predictions to output file in CSV format
    with open(output_file, "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["user_id", "business_id", "prediction"])
        for row in predictions:
            writer.writerow(row)
    
    # ---------------------------
    # 7. Optional RMSE Calculation
    # ---------------------------
    # Try to calculate RMSE if validation data has ratings (for self-evaluation)
    try:
        # Read validation data with ratings (same file but including the rating column)
        validation_rdd = read_test_data(test_file, include_ratings=True)
        
        # Only proceed if we have validation data
        if validation_rdd.count() > 0:
            # Calculate and print RMSE
            rmse = calculate_rmse(predictions, validation_rdd)
            print(f"RMSE: {rmse}")
            
            # ---------------------------
            # 8. Error Distribution Analysis
            # ---------------------------
            # Create a dictionary mapping (user_id, business_id) to actual rating
            validation_dict = validation_rdd.map(lambda x: ((x[0], x[1]), x[2])).collectAsMap()
            
            # Initialize counters for different error ranges
            error_ranges = {">=0 and <1": 0, ">=1 and <2": 0, ">=2 and <3": 0, ">=3 and <4": 0, ">=4": 0}
            
            # Count number of predictions in each error range
            for user_id, business_id, pred_rating in predictions:
                key = (user_id, business_id)
                if key in validation_dict:
                    actual = validation_dict[key]
                    error = abs(pred_rating - actual)
                    
                    # Categorize error into appropriate range
                    if error < 1:
                        error_ranges[">=0 and <1"] += 1
                    elif error < 2:
                        error_ranges[">=1 and <2"] += 1
                    elif error < 3:
                        error_ranges[">=2 and <3"] += 1
                    elif error < 4:
                        error_ranges[">=3 and <4"] += 1
                    else:
                        error_ranges[">=4"] += 1
            
            # Print error distribution
            print("Error Distribution:")
            for range_label, count in error_ranges.items():
                print(f"  {range_label}: {count}")
    except:
        # If RMSE calculation fails, just skip it
        # This keeps the output format intact for grading
        pass
    
    # ---------------------------
    # 9. Finalization
    # ---------------------------
    # Calculate and print execution time
    execution_time = time.time() - start_time
    print(f"Task 2.1 completed successfully in {execution_time:.2f} seconds.")
    
    # Clean up by stopping the Spark context
    sc.stop()

if __name__ == "__main__":
    main()